retrofit/src/main/java/retrofit2/HttpServiceMethod.java

```diff
@@ -74,20 +98,41 @@
 
   private final RequestFactory requestFactory;
   private final okhttp3.Call.Factory callFactory;
-  private final CallAdapter<ResponseT, ReturnT> callAdapter;
+  /** Null indicates a Kotlin coroutine service method. */
+  private final @Nullable CallAdapter<ResponseT, ReturnT> callAdapter;
+  /**
+   * True if the coroutine continuation should receive the full Response object. Only meaningful
+   * when {@link #callAdapter} is null.
+   */
+  private final boolean continuationWantsResponse;
   private final Converter<ResponseBody, ResponseT> responseConverter;
 
-  private HttpServiceMethod(RequestFactory requestFactory, okhttp3.Call.Factory callFactory,
-      CallAdapter<ResponseT, ReturnT> callAdapter,
+  private HttpServiceMethod(RequestFactory requestFactory, Call.Factory callFactory,
+      @Nullable CallAdapter<ResponseT, ReturnT> callAdapter, boolean continuationWantsResponse,
       Converter<ResponseBody, ResponseT> responseConverter) {
     this.requestFactory = requestFactory;
     this.callFactory = callFactory;
     this.callAdapter = callAdapter;
+    this.continuationWantsResponse = continuationWantsResponse;
     this.responseConverter = responseConverter;
   }
 
   @Override ReturnT invoke(@Nullable Object[] args) {
-    return callAdapter.adapt(
-        new OkHttpCall<>(requestFactory, args, callFactory, responseConverter));
+    OkHttpCall<ResponseT> call =
+        new OkHttpCall<>(requestFactory, args, callFactory, responseConverter);
+
+    if (callAdapter != null) {
+      return callAdapter.adapt(call);
+    }
+
+    //noinspection ConstantConditions Suspension functions always have arguments.
+    Object continuation = args[args.length - 1];
+    if (continuationWantsResponse) {
+      //noinspection unchecked Guaranteed by parseAnnotations above.
+      return (ReturnT) KotlinExtensions.awaitResponse(call,
```

Comment:
This function name looks like it’ll be blocking but my assumption is that it'll return a promise. I gotta read more.

Reply:
It actually returns a union type of `COROUTINE_SUSPENDED | T` which is either a marker value indication suspension to unwind the stack or the result if it was able to be computed synchronously. Since we're invoking `Call.enqueue` it will always return `COROUTINE_SUSPENDED`.